### PR TITLE
Bluetooth: has: Make HAS registration dynamic

### PR DESCRIPTION
**Pull request by larsgk**
This changes HAS registration to be dynamic
and let's the application set Hearing Aid Type

Often, devices are flashed with generic firmware with some
features stored post factory production, requiring
the settings to be moved from compile time to run-time.
    
This change will increase the RAM usage as the GATT
service is moved from ROM to RAM.

Signed-off-by: Lars Knudsen <larsgk@gmail.com>
Co-author: Soren Engquist <soren@engquist.dk>

### CODE REVIEW

File: include/zephyr/bluetooth/audio/has.h [hunk 1]
```diff
@@ -67,6 +67,12 @@ enum bt_has_capabilities {
 	BT_HAS_PRESET_SUPPORT = BIT(0),
 };
 
+/** @brief Structure for registering a Hearing Access Service instance. */
+struct bt_has_register_param {
+	/** Hearing Aid Type value */
+	uint8_t hearing_aid_type;
```
Comment (MariuszSkamra) [inline include/zephyr/bluetooth/audio/has.h:73]:
```suggestion
	enum bt_has_hearing_aid_type type;
```

File: subsys/bluetooth/audio/has.c [hunk 2]
```diff
@@ -70,33 +71,42 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
...
-	BT_GATT_PRIMARY_SERVICE(BT_UUID_HAS),
-	BT_AUDIO_CHRC(BT_UUID_HAS_HEARING_AID_FEATURES,
-		      BT_GATT_CHRC_READ,
-		      BT_GATT_PERM_READ_ENCRYPT,
-		      read_features, NULL, NULL),
-#if defined(CONFIG_BT_HAS_PRESET_SUPPORT)
-	BT_AUDIO_CHRC(BT_UUID_HAS_PRESET_CONTROL_POINT,
 #if defined(CONFIG_BT_EATT)
-		      BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE | BT_GATT_CHRC_NOTIFY,
+#define CONTROL_POINT_METHODS (BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE | BT_GATT_CHRC_NOTIFY)
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/has.c:75]:
```suggestion
#define CONTROL_POINT_PROP (BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE | BT_GATT_CHRC_NOTIFY)
```

It's GATT Characteristic Properties bitfield as defined in Core

File: subsys/bluetooth/audio/has.c [hunk 3]
```diff
@@ -1122,12 +1132,23 @@ int bt_has_preset_name_change(uint8_t index, const char *name)
 }
 #endif /* CONFIG_BT_HAS_PRESET_SUPPORT */
 
-static int has_init(const struct device *dev)
+int bt_has_register(const struct bt_has_register_param *param)
 {
-	ARG_UNUSED(dev);
+	LOG_DBG("Trying to start HAS service...");
+	int err;
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/has.c:1138]: The log here is redundant as it does not give any additional information.
You could simply put `LOG_DBG("")` instead which will print `bt_has_register` function name.
I would suggest to at least log the `param` pointer. Variable declaration first.

```suggestion
	int err;
	
	LOG_DBG("param %p", param);
```

File: subsys/bluetooth/audio/has.c [hunk 4]
```diff
@@ -70,33 +71,42 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
...
+
+#define BT_HAS_DEFINITION \
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_HAS), \
+	BT_AUDIO_CHRC(BT_UUID_HAS_HEARING_AID_FEATURES, \
+		      BT_GATT_CHRC_READ, \
+		      BT_GATT_PERM_READ_ENCRYPT, \
+		      read_features, NULL, NULL), \
+	BT_HAS_PRESET_ATTRS
+
+static struct bt_gatt_attr has_attrs[] = { BT_HAS_DEFINITION };
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/has.c:104]:
```suggestion
static struct bt_gatt_attr has_attrs[] = {
	BT_GATT_PRIMARY_SERVICE(BT_UUID_HAS),
	BT_AUDIO_CHRC(BT_UUID_HAS_HEARING_AID_FEATURES,
		      BT_GATT_CHRC_READ,
		      BT_GATT_PERM_READ_ENCRYPT,
		      read_features, NULL, NULL),
#if defined(CONFIG_BT_HAS_PRESET_SUPPORT)
	BT_AUDIO_CHRC(BT_UUID_HAS_PRESET_CONTROL_POINT,
		      CONTROL_POINT_METHODS,
		      BT_GATT_PERM_WRITE_ENCRYPT,
		      NULL, write_control_point, NULL),
	BT_AUDIO_CCC(ccc_cfg_changed),
	BT_AUDIO_CHRC(BT_UUID_HAS_ACTIVE_PRESET_INDEX,
		      BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
		      BT_GATT_PERM_READ_ENCRYPT,
		      read_active_preset_index, NULL, NULL),
	BT_AUDIO_CCC(ccc_cfg_changed),
#endif /* CONFIG_BT_HAS_PRESET_SUPPORT */
};
```

Looks cleaner IMO

**Review by MariuszSkamra — CHANGES_REQUESTED** [review]
Few changes requested, otherwise LGTM

File: subsys/bluetooth/audio/has.c (hunk 3 above)
Comment (larsgk) [reply subsys/bluetooth/audio/has.c:1138]: That should just have been removed - missed it

File: subsys/bluetooth/audio/has.c (hunk 4 above)
Comment (larsgk) [reply subsys/bluetooth/audio/has.c:104]: you are right, I'll revert the structure and add it directly in has_attrs

File: subsys/bluetooth/audio/has.c (hunk 2 above)
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/has.c:75]:
```suggestion
#define CONTROL_POINT_PROPERTIES (BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE | BT_GATT_CHRC_NOTIFY)
```

I would prefer a naming such as this, as they actually are properties in spec., and it also the name of the argument in BT_AUDIO_CCC where we use this

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c [hunk 5]
```diff
@@ -425,6 +425,25 @@ int bap_unicast_sr_init(void)
 
 	bt_pacs_cap_register(BT_AUDIO_DIR_SINK, &cap_sink);
 
+	if (IS_ENABLED(CONFIG_BT_PAC_SNK_LOC)) {
+		if (IS_ENABLED(CONFIG_BT_HAS_HEARING_AID_BANDED)) {
```
[the remark this answers is not in the document]
Comment (larsgk) [reply samples/bluetooth/hap_ha/src/bap_unicast_sr.c:429]: I kept them for now - but maybe we should change them to be enums in has.h?

File: subsys/bluetooth/audio/has.c [hunk 6]
```diff
@@ -1122,12 +1132,23 @@ int bt_has_preset_name_change(uint8_t index, const char *name)
...
 
-static int has_init(const struct device *dev)
+int bt_has_register(const struct bt_has_register_param *param)
 {
-	ARG_UNUSED(dev);
+	LOG_DBG("Trying to start HAS service...");
+	int err;
+
+	CHECKIF(!param) {
+		LOG_DBG("NULL params pointer");
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/has.c:1141]: I think that we should use BT_DBG here instead

File: subsys/bluetooth/audio/has.c [hunk 7]
```diff
@@ -1122,12 +1132,23 @@ int bt_has_preset_name_change(uint8_t index, const char *name)
...
+		return -EINVAL;
+	}
+
+	CHECKIF(param->hearing_aid_type > 2) {
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/has.c:1145]: I'm not a fan of this magic number :)
I would assume that BT_HAS_HEARING_AID_TYPE_BANDED is the max of enum, and its very unlikely to change. Check this one instead, or create a BT_HAS_HEARING_AID_TYPE_INVALID as max and check?

File: subsys/bluetooth/audio/has.c [hunk 8]
```diff
@@ -1122,12 +1132,23 @@ int bt_has_preset_name_change(uint8_t index, const char *name)
...
+		LOG_DBG("Invalid Hearing Aid Type 0x%x", param->hearing_aid_type);
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/has.c:1146]: BT_DBG

File: subsys/bluetooth/audio/has.c (hunk 2 above)
Comment (fredrikdanebjer) [reply subsys/bluetooth/audio/has.c:75]: I see that Mariusz already commented this :)

File: subsys/bluetooth/audio/has.c (hunk 8 above)
Comment (Thalley) [reply subsys/bluetooth/audio/has.c:1146]: @fredrikdanebjer The BT_DBG/WARN/ERR logging functions has been removed by https://github.com/zephyrproject-rtos/zephyr/pull/51897 :)

**Comment by Thalley** [comment]
@larsgk Will review after UPF :)

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply samples/bluetooth/hap_ha/src/bap_unicast_sr.c:429]: It would be nice as it's dead code actually (not used by stack). Leaving it might be confusing to other users

**Review by MariuszSkamra — CHANGES_REQUESTED** [review]
As the `bt_has_hearing_aid_type` represent 2 LSB bits in features, you need to redefine the enum values so that:

```
enum bt_has_hearing_aid_type {
	BT_HAS_HEARING_AID_TYPE_BINAURAL = BIT(0),
	BT_HAS_HEARING_AID_TYPE_MONAURAL = BIT(1),
	BT_HAS_HEARING_AID_TYPE_BANDED = BIT(2),
};
```

File: subsys/bluetooth/audio/has.c (hunk 7 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/has.c]: or just check all variants, as there are only 3 possibilities.

File: subsys/bluetooth/audio/has.c (hunk 6 above)
Comment (Thalley) [reply subsys/bluetooth/audio/has.c:1135]: See https://github.com/zephyrproject-rtos/zephyr/pull/52672?notification_referrer_id=NT_kwDOACZ1XbI0OTYyMjkzMzc4OjI1MjA0MTM#discussion_r1036895465

**Comment by larsgk** [comment]
> As the `bt_has_hearing_aid_type` represent 2 LSB bits in features, you need to redefine the enum values so that:
> 
> ```
> enum bt_has_hearing_aid_type {
> 	BT_HAS_HEARING_AID_TYPE_BINAURAL = BIT(0),
> 	BT_HAS_HEARING_AID_TYPE_MONAURAL = BIT(1),
> 	BT_HAS_HEARING_AID_TYPE_BANDED = BIT(2),
> };
> ```

oops - will fix ;)

File: subsys/bluetooth/audio/has.c (hunk 7 above)
Comment (larsgk) [reply subsys/bluetooth/audio/has.c]: > If `hearing_aid_type` is a an enum, we could/should consider not doing this check, right?

true - fixing

File: samples/bluetooth/hap_ha/src/bap_unicast_sr.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (larsgk) [reply samples/bluetooth/hap_ha/src/bap_unicast_sr.c:429]: I moved the configs to the sample - making it possible to (still) configure the type

File: subsys/bluetooth/shell/has.c [hunk 9]
```diff
@@ -75,6 +75,35 @@ static int cmd_preset_unreg(const struct shell *sh, size_t argc, char **argv)
...
+	for (size_t argn = 1; argn < argc; argn++) {
+		const char *arg = argv[argn];
+
+		if (strcmp(arg, "binaural") == 0) {
+			param.type = BT_HAS_HEARING_AID_TYPE_BINAURAL;
+		} else if (strcmp(arg, "banded") == 0) {
+			param.type = BT_HAS_HEARING_AID_TYPE_BANDED;
+		} else {
+			shell_help(sh);
+			return SHELL_CMD_HELP_PRINTED;
```
Comment (MariuszSkamra) [inline subsys/bluetooth/shell/has.c:94]: You'd still need to handle `monaural` argument, even if it's default

File: subsys/bluetooth/audio/Kconfig.has [hunk 10]
```diff
@@ -70,7 +27,9 @@ config BT_HAS_PRESET_SUPPORT
 
 if BT_HAS_PRESET_SUPPORT
 
-if BT_HAS_HEARING_AID_BINAURAL
+# Note: BT_HAS_IDENTICAL_PRESET_RECORDS & BT_HAS_PRESET_SYNC_SUPPORT below
+# are only relevant for binaural hearing aids.
+# TODO: Move from Kconfig to bt_has_register_param in a subsequent PR
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/Kconfig.has:32]: Apart form the note, I think it's good to ignore those Kconfig option values in `has.c` if non-binaural type is registered

**Review by MariuszSkamra — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/shell/has.c (hunk 9 above)
Comment (larsgk) [reply subsys/bluetooth/shell/has.c:94]: true 🤦‍♂️

**Review by MariuszSkamra — DISMISSED** [review]
LGTM

File: subsys/bluetooth/audio/has.c [hunk 11]
```diff
@@ -1147,23 +1157,11 @@ static int has_init(const struct device *dev)
...
-					     (BT_AUDIO_LOCATION_FRONT_LEFT |
-					      BT_AUDIO_LOCATION_FRONT_RIGHT));
-		} else if (IS_ENABLED(CONFIG_BT_HAS_HEARING_AID_LEFT)) {
-			bt_pacs_set_location(BT_AUDIO_DIR_SINK,
-					     BT_AUDIO_LOCATION_FRONT_LEFT);
-		} else {
-			bt_pacs_set_location(BT_AUDIO_DIR_SINK,
-					     BT_AUDIO_LOCATION_FRONT_RIGHT);
-		}
+	has_svc = (struct bt_gatt_service)BT_GATT_SERVICE(has_attrs);
```
[the remark this answers is not in the document]
Comment (fredrikdanebjer) [reply subsys/bluetooth/audio/has.c:1160]: I agree :)

[the remark this answers is not in the document]
Comment (fredrikdanebjer) [reply subsys/bluetooth/audio/has.c:1160]: I retract this comment. Seems to me like this is the style mostly used in audio so far, I'd like to keep it as is.

**Review by fredrikdanebjer — DISMISSED** [review]

File: include/zephyr/bluetooth/audio/has.h [hunk 12]
```diff
@@ -45,9 +45,9 @@ struct bt_has;
 
 /** Hearing Aid device type */
 enum bt_has_hearing_aid_type {
-	BT_HAS_HEARING_AID_TYPE_BINAURAL,
-	BT_HAS_HEARING_AID_TYPE_MONAURAL,
-	BT_HAS_HEARING_AID_TYPE_BANDED,
+	BT_HAS_HEARING_AID_TYPE_BINAURAL = BIT(0),
+	BT_HAS_HEARING_AID_TYPE_MONAURAL = BIT(1),
+	BT_HAS_HEARING_AID_TYPE_BANDED = BIT(2),
```
Comment (Thalley) [inline include/zephyr/bluetooth/audio/has.h:50]: Any reason why we are using `BIT(x)` instead of just `0x0X`? 

IMO using `BIT(X)` makes it seem like it is a bitfield (which it isn't).

File: samples/bluetooth/hap_ha/Kconfig [hunk 13]
```diff
@@ -19,4 +19,34 @@ config HAP_HA_SET_RANK
...
+	help
+	  Select the Hearing Aid Device location.
+
+config BT_HAS_HEARING_AID_LEFT
+	bool "Left Ear"
+
+config BT_HAS_HEARING_AID_RIGHT
+	bool "Right Ear"
+
+endchoice # BT_HAS_HEARING_AID_LOCATION
```
Comment (Thalley) [inline samples/bluetooth/hap_ha/Kconfig:50]: Since this is in a sample, I suggest we remove the `BT_HAS` and `BT` prefix, as they could otherwise clash with BT subsystem Kconfigs.

File: subsys/bluetooth/audio/Kconfig.has (hunk 10 above)
Comment (Thalley) [inline subsys/bluetooth/audio/Kconfig.has:32]: What is the reasoning for not adding them in this PR? Sounds like a fairly small change, and between the two PRs, HAS would otherwise be left in a kind of weird state.

File: subsys/bluetooth/shell/has.c [hunk 14]
```diff
@@ -216,6 +247,10 @@ static int cmd_has(const struct shell *sh, size_t argc, char **argv)
 }
 
 SHELL_STATIC_SUBCMD_SET_CREATE(has_cmds,
+	SHELL_CMD_ARG(register, NULL,
+		      "Initialize the service and register type "
+		      "[binaural|monaural(default)|banded]",
```
Comment (Thalley) [inline subsys/bluetooth/shell/has.c:252]:
```suggestion
		      "[binaural | monaural(default) | banded]",
```

**Review by Thalley — CHANGES_REQUESTED** [review]
Please also add an explanation as to why this change is being done in your PR/commit description. 

We've discussed this on Discord, and while I generally disagree with this approach, I understand why it is being implemented like this, but please add that explanation here.

File: samples/bluetooth/hap_ha/Kconfig (hunk 13 above)
Comment (larsgk) [reply samples/bluetooth/hap_ha/Kconfig:50]: Makes sense - I will change the prefix to the HAP_HA_* used above

File: include/zephyr/bluetooth/audio/has.h (hunk 12 above)
Comment (larsgk) [reply include/zephyr/bluetooth/audio/has.h:50]: true - it's a bug (will change back to simple enum)

File: subsys/bluetooth/audio/has.c [hunk 15]
```diff
@@ -1122,48 +1125,51 @@ int bt_has_preset_name_change(uint8_t index, const char *name)
...
-		if (!IS_ENABLED(CONFIG_BT_HAS_IDENTICAL_PRESET_RECORDS)) {
-			has.features |= BT_HAS_FEAT_INDEPENDENT_PRESETS;
+		has.features |= BT_HAS_FEAT_PRESET_SYNC_SUPP;
+	}
+
+	if (param->independent_presets) {
+		if (param->type != BT_HAS_HEARING_AID_TYPE_BINAURAL) {
+			LOG_DBG("Independent presets only available for binaural hearing aid type");
+			return -EINVAL;
 		}
```
Comment (Thalley) [inline subsys/bluetooth/audio/has.c:1159]: The documentation for the `struct bt_has_register_param` should contain these restrictions :)

File: subsys/bluetooth/audio/has.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/audio/has.c:1168]: No strong opinions from me.

File: subsys/bluetooth/shell/has.c [hunk 16]
```diff
@@ -75,6 +75,43 @@ static int cmd_preset_unreg(const struct shell *sh, size_t argc, char **argv)
...
+
+	for (size_t argn = 1; argn < argc; argn++) {
+		const char *arg = argv[argn];
+
+		if (strcmp(arg, "binaural") == 0) {
+			param.type = BT_HAS_HEARING_AID_TYPE_BINAURAL;
+		} else if (strcmp(arg, "monaural") == 0) {
+			param.type = BT_HAS_HEARING_AID_TYPE_MONAURAL;
+		} else if (strcmp(arg, "banded") == 0) {
+			param.type = BT_HAS_HEARING_AID_TYPE_BANDED;
```
Comment (Thalley) [inline subsys/bluetooth/shell/has.c:95]: Since these values are basically a choice, rather than 3 individual parameters, I would suggest to keep the hearing aid type as a mandatory argument to `has register` and leave the `sync` and `independent` as optional parameters. 

Basically changing

```
"[binaural | monaural(default) | banded] [sync] [independent]",
cmd_has_register, 1, 3),
```
to 
```
"<binaural | monaural | banded> [sync] [independent]",
cmd_has_register, 2, 2),
```

Does that make sense?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: include/zephyr/bluetooth/audio/has.h [hunk 17]
```diff
@@ -45,7 +45,7 @@ struct bt_has;
 
 /** Hearing Aid device type */
 enum bt_has_hearing_aid_type {
-	BT_HAS_HEARING_AID_TYPE_BINAURAL,
+	BT_HAS_HEARING_AID_TYPE_BINAURAL = 0,
 	BT_HAS_HEARING_AID_TYPE_MONAURAL,
 	BT_HAS_HEARING_AID_TYPE_BANDED,
```
[the remark this answers is not in the document]
Comment (larsgk) [reply include/zephyr/bluetooth/audio/has.h:50]: > We could consider making them with 0x01, 0x02 and 0x03 as those are the literal values that the HAS spec defines, and that we can be more sure that they are always correct

I would then rather use the BIT notation that is also used just below  0, BIT(0), BIT(1)

File: subsys/bluetooth/audio/Kconfig.has [hunk 18]
```diff
@@ -70,24 +27,6 @@ config BT_HAS_PRESET_SUPPORT
...
-
-config BT_HAS_PRESET_SYNC_SUPPORT
-	bool "Preset synchronization support"
-	depends on BT_HAS_IDENTICAL_PRESET_RECORDS
-	help
-	  Set if the hearing aid has support for relaying active preset changes to the other
-	  member in the Binaural Hearing Aid Set.
-
-endif # BT_HAS_HEARING_AID_BINAURAL
-
```
[the remark this answers is not in the document]
Comment (larsgk) [reply subsys/bluetooth/audio/Kconfig.has:90]: > Since the setting of these values are now moved to the register par, should these be moved to, and used in, the sample Kconfig and implementation?

It could make sense, yes.  I would then rather do it together with adding something around it for the BSIM test in a subsequent commit.

File: subsys/bluetooth/audio/has.c (hunk 15 above)
Comment (larsgk) [reply subsys/bluetooth/audio/has.c:1159]: > The documentation for the `struct bt_has_register_param` should contain these restrictions :)

You are right - forgot it ends up in the actual documentation ;)

File: subsys/bluetooth/shell/has.c (hunk 16 above)
Comment (larsgk) [reply subsys/bluetooth/shell/has.c:95]: I like to have a default for convenience - something that works on the minimal setup with just one device. Maybe @fredrikdanebjer or @MariuszSkamra has an opinion on this?

File: subsys/bluetooth/audio/has.c (hunk 15 above)
Comment (larsgk) [reply subsys/bluetooth/audio/has.c:1159]: I added a remark on this being for binaural only

File: include/zephyr/bluetooth/audio/has.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply include/zephyr/bluetooth/audio/has.h]: I understand why you see that as a good option, since it's what the spec uses. My opinion is still that the way the spec presents it, using bits and not values, is a poor choice as it is not a bitfield. For that same reason, I believe that 

```c
	BT_HAS_HEARING_AID_TYPE_BINAURAL = 0x00,
	BT_HAS_HEARING_AID_TYPE_MONAURAL = 0x01,
	BT_HAS_HEARING_AID_TYPE_BANDED = 0x02,
```

is the best option, to avoid any confusion. 

@MariuszSkamra any input?

[the remark this answers is not in the document]
Comment (larsgk) [reply include/zephyr/bluetooth/audio/has.h]: @Thalley - I see your point.  I'll change it to hex values for clarity.

**Review by MariuszSkamra — DISMISSED** [review]

File: include/zephyr/bluetooth/audio/has.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply include/zephyr/bluetooth/audio/has.h]: > It is a bitfield actually, as the Hearing Aid type are two LSB in Hearing Aid Features characteristic. I won't object if you leave it as hex, it's just a matter of preference

True, saying it isn't a bitfield is wrong on my side. It is a bitfield, but where the bits are exclusive, i.e. you cannot set more than one bit. The "cannot set more than one bit" is what I want to empathize, which I am worried that using the `BIT(x)` macro won't. 

We can keep the `BIT(x)` to stay more "true" to the spec, if we add proper documentation of this, and ensuring that no more than 1 bit is set. 

`0xb11` is currently RFU, but if they add another type `BT_HAS_HEARING_AID_TYPE_NEW`, which would have the value `0b11`, then suddenly (if using `BIT(x)`) we might indicate that `BT_HAS_HEARING_AID_TYPE_MONAURAL | BT_HAS_HEARING_AID_TYPE_MONAURAL == BT_HAS_HEARING_AID_TYPE_NEW`, because the `OR`/`|` is typically something you can do with values defined with `BIT(x)`. 

TL;DR: I am OK with either hex values (0x00) or using the `BIT(x)` (although we cannot denote `BT_HAS_HEARING_AID_TYPE_BINAURAL` with `BIT(x)`) if it gets properly documented, though I still prefer the hex values.

**Review by Thalley — APPROVED** [review]

**Review by Thalley — CHANGES_REQUESTED** [review]
Almost forgot: Please update the commit and PR message to explain why this change is being done as requested earlier: https://github.com/zephyrproject-rtos/zephyr/pull/52672/#pullrequestreview-1215165084

**Review by Thalley — APPROVED** [review]
If we truly want to make HAS dynamically registered, this is OK as it is for me now :)

File: subsys/bluetooth/shell/has.c [hunk 19]
```diff
@@ -75,6 +75,43 @@ static int cmd_preset_unreg(const struct shell *sh, size_t argc, char **argv)
...
+	struct bt_has_register_param param = {
+		.type = BT_HAS_HEARING_AID_TYPE_MONAURAL,
+		.preset_sync_support = false,
+		.independent_presets = false
+	};
+
+	for (size_t argn = 1; argn < argc; argn++) {
+		const char *arg = argv[argn];
+
+		if (strcmp(arg, "binaural") == 0) {
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/shell/has.c:90]:
```suggestion
		if (strcmp(argv[1], "binaural") == 0) {
```
Sorry, a small detail, but I would prefer it close to this (and adjust the rest accordingly), to avoid several type inputs. This seems also like how its been done in other places for shell

Comment (fredrikdanebjer) [reply subsys/bluetooth/shell/has.c:90]: Seems there are both approaches already. Personally Id prefer locked positions for the arguments, but I dont think it should stand in the way of this now.

**Review by fredrikdanebjer — APPROVED** [review]
